haystack/components/agents/agent.py

```diff
@@ -610,14 +550,25 @@ def run(  # noqa: PLR0915
                             },
                             component_visits=exe_context.component_visits,
                             parent_span=span,
+                            break_point=break_point.break_point if isinstance(break_point, AgentBreakpoint) else None,
                         )
-                    except PipelineRuntimeError as e:
-                        pipeline_snapshot = _create_pipeline_snapshot_from_chat_generator(
-                            agent_name=getattr(self, "__component_name__", None),
-                            execution_context=exe_context,
-                            parent_snapshot=parent_snapshot,
+                    except (BreakpointException, PipelineRuntimeError) as e:
+                        if isinstance(e, BreakpointException):
```

Comment:
This code, which handles the chat generator exception, is very similar to the code that handles the tool invoker exception. A helper function can be made to avoid this code duplication.

Also, one calls the:`_create_pipeline_snapshot_from_chat_generator()`
while the other the `_create_pipeline_snapshot_from_tool_invoker()` 

But maybe those two can be made into a single function with a `component_type` parameter then this two blocks of code to handle the chat_generator/tool_invocation exception could be the same.

Reply:
I tried to merge the two functions together, but in the end the new function was about the same length as the two separate ones due to all the subtle differences between the two. So I'll opt to keep both separate for sake of clarity.